packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -83,6 +94,15 @@ export function serveRawFsMiddleware(): Connect.NextHandleFunction {
     // the paths are rewritten to `/@fs/` prefixed paths and must be served by
     // searching based from fs root.
     if (url.startsWith(FS_PREFIX)) {
+      // restrict files outside of `fsServeRoot`
+      if (
+        normalizePath(path.relative(root, fsPathFromId(url))).startsWith('../')
```

Comment:
Honestly - to me this is horrifying. But sadly also the more or less the solution I found in stackoverflow. And even more sadly node.js doesn't seem to have a nice function for that although that should be a quite common thing.

I'm on my phone so I can't check right now and this is probably a correct implementation but I really don't like checking for forbidden paths for such critical code. I would rather like if it would check for allowed paths. An idea that *may* work would be to get the normalized absolute paths and check if the target one starts with the normalized absolute `root` path. There are also functions in the path module for that.

Reply:
I think it could be written like `path.resolve( fsPathFromId(url) ).startsWith(config.root + '/'))` that is something we use in other places, but this should be equivalent. @mohe2015 if you have a better way to do this, please propose a changeset, or even if this PR gets merged, you could send a new PR to improve it later

Reply:
It would be great if you could do that change here if that's possible.

Reply:
Of config.root is not normalized before this should probably also be done

Reply:
I'm not sure if this would work on windows (which has `\` path separators), maybe normalizePath would fix that. I would recommend adding some unit tests for the edge cases.
I would really recommend @patak-js 's `startsWith()` solution, which is more solid. It avoids complexity like
```
Request variations
Encoding and double encoding:

%2e%2e%2f represents ../
%2e%2e/ represents ../
..%2f represents ../ 
%2e%2e%5c represents ..\
%2e%2e\ represents ..\ 
..%5c represents ..\ 
%252e%252e%255c represents ..\ 
..%255c represents ..\ 
and so on.

Percent encoding (aka URL encoding)
Note that web containers perform one level of decoding on percent encoded values from forms and URLs.

..%c0%af represents ../ 
..%c1%9c represents ..\ 
```

https://owasp.org/www-community/attacks/Path_Traversal

Reply:
Is it simply this check?:

```js
// normalizePath(path.relative(root, fsPathFromId(url))).startsWith('../')
/^\.\.[\\/].+/.test(normalizePath(path.relative(root, fsPathFromId(url))))
```

Or is it more complicated?

Reply:
> Is it simply this check?:
> 
> ```js
> // normalizePath(path.relative(root, fsPathFromId(url))).startsWith('../')
> /^\.\.[\\/].+/.test(normalizePath(path.relative(root, fsPathFromId(url))))
> ```
> 
> Or is it more complicated?

Just use @patak-js 's solution it's so much more secure.

Reply:
Agree with @mohe2015, it's generally much safer to normalize and resolve the full absolute path. Then we can check if it resolves to one of the allowed directories to serve from. Requires less logic and we can reuse more platform APIs.